File: perl5/sync_agent_run.pl
#!/usr/bin/perl
# sync_agent_run.py — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:scripts/sync_agent_run.py
# auch in: OpenClaw@gateway2:scripts/sync_agent_run.py
# Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use lib '/home/openclaw/.openclaw/workspace/scripts';
use File::Find;
use File::stat;
use JSON;
use Time::Piece;

# Lade Funktionen aus externem Modul
require "sync_clawhub_git.pl";

my $CLAWHUB_DIR = '/home/openclaw/.openclaw/workspace/skills';
my $GIT_DIR = '/home/openclaw/.openclaw/workspace/git/skills';

sub file_mtime {
    my ($path) = @_;
    my @files;
    find(sub {
        return if -d $_ || $_ eq '.git';
        push @files, $File::Find::name;
    }, $path);
    
    return 0 unless @files;
    
    my $max_time = 0;
    for my $file (@files) {
        my $mtime = (stat($file))[9];
        $max_time = $mtime if $mtime > $max_time;
    }
    return $max_time;
}

sub log_message {
    my ($message, $level) = @_;
    $level //= "INFO";
    print "$message\n";
}

log_message("=" x 70);
log_message("CLAWHUB ↔ GIT SYNC AGENT - PRODUKTIONS-LAUF");
log_message("Zeitstempel: " . localtime->datetime);
log_message("=" x 70);

sub get_directories {
    my ($base_dir) = @_;
    opendir(my $dh, $base_dir) or return ();
    my @dirs = grep { !/^\./ && -d "$base_dir/$_" } readdir($dh);
    closedir $dh;
    return @dirs;
}

my @clawhub_skills = get_directories($CLAWHUB_DIR);
my @git_skills = get_directories($GIT_DIR);

my %clawhub_hash = map { $_ => 1 } @clawhub_skills;
my %git_hash = map { $_ => 1 } @git_skills;

my %results = (
    synced_to_git => [],
    synced_to_clawhub => [],
    up_to_date => [],
    errors => []
);

# Phase 1: Neu in ClawHub -> zu Git syncen
log_message("\n[PHASE 1] ClawHub → Git Synchronisation");
log_message("-" x 40);

my @new_in_clawhub = sort grep { !$git_hash{$_} } @clawhub_skills;
for my $skill (@new_in_clawhub) {
    eval {
        my $skill_path = "$CLAWHUB_DIR/$skill";
        if (validate_skill($skill_path)) {
            log_message("→ Synchronisiere $skill zu Git...");
            if (sync_to_git($skill, 0)) {  # 0 = nicht dry-run
                my $git_path = "$GIT_DIR/$skill";
                chdir($git_path);
                system('git init -q 2>/dev/null');
                system('git add . -f 2>/dev/null');
                my $dt = localtime->strftime("%Y-%m-%d %H:%M");
                system("git commit -m \"Initial: $skill\" -q 2>/dev/null");
                push @{$results{synced_to_git}}, $skill;
                log_message("  ✓ $skill synchronisiert & Git initialisiert");
            } else {
                push @{$results{errors}}, "$skill (sync failed)";
            }
        } else {
            push @{$results{errors}}, "$skill (invalid)";
        }
        1;
    } or do {
        my $error = $@ || 'Unknown error';
        log_message("  ✗ ERROR: $skill - $error", "ERROR");
        push @{$results{errors}}, "$skill (exception)";
    };
}

# Phase 2: Prüfe existierende Skills auf Änderungen
log_message("\n[PHASE 2] Prüfe existierende Skills auf Änderungen");
log_message("-" x 40);

my @in_both = sort grep { $git_hash{$_} } @clawhub_skills;
for my $skill (@in_both) {
    eval {
        my $c_mtime = file_mtime("$CLAWHUB_DIR/$skill");
        my $g_mtime = file_mtime("$GIT_DIR/$skill");
        my $diff = $c_mtime - $g_mtime;

        if (abs($diff) > 60) {
            if ($diff > 0) {
                log_message("→ $skill: ClawHub neuer (+${diff}s) → sync zu Git");
                if (sync_to_git($skill, 0)) {
                    my $git_path = "$GIT_DIR/$skill";
                    chdir($git_path);
                    system('git add . -f 2>/dev/null');
                    my $dt = localtime->strftime("%Y-%m-%d %H:%M");
                    system("git commit -m \"Sync from ClawHub: $dt\" -q 2>/dev/null");
                    push @{$results{synced_to_git}}, $skill;
                } else {
                    push @{$results{errors}}, "$skill (update failed)";
                }
            } else {
                log_message("→ $skill: Git neuer (+".abs($diff)."s) → sync zu ClawHub");
                if (sync_to_clawhub($skill, 0)) {
                    push @{$results{synced_to_clawhub}}, $skill;
                } else {
                    push @{$results{errors}}, "$skill (update failed)";
                }
            }
        } else {
            push @{$results{up_to_date}}, $skill;
        }
        1;
    } or do {
        my $error = $@ || 'Unknown error';
        log_message("  ✗ ERROR: $skill - $error", "ERROR");
        push @{$results{errors}}, "$skill (exception)";
    };
}

# Zusammenfassung
log_message("\n" . "=" x 70);
log_message("SYNCHRONISATION ABGESCHLOSSEN");
log_message("=" x 70);
log_message(sprintf("Zu Git synchronisiert:     %d", scalar @{$results{synced_to_git}}));
if (@{$results{synced_to_git}}) {
    log_message("  " . join(", ", @{$results{synced_to_git}}));
}
log_message(sprintf("Zu ClawHub synchronisiert: %d", scalar @{$results{synced_to_clawhub}}));
if (@{$results{synced_to_clawhub}}) {
    log_message("  " . join(", ", @{$results{synced_to_clawhub}}));
}
log_message(sprintf("Bereits aktuell:           %d", scalar @{$results{up_to_date}}));
log_message(sprintf("Fehler:                    %d", scalar @{$results{errors}}));
if (@{$results{errors}}) {
    log_message("  " . join(", ", @{$results{errors}}));
}
log_message("=" x 70);

# Speichere State
my $STATE_FILE = "/home/openclaw/.openclaw/workspace/db/sync_state.json";
my $state_dir = "/home/openclaw/.openclaw/workspace/db";
mkdir $state_dir unless -d $state_dir;

my $state = {
    last_run => localtime->datetime,
    results => \%results
};

open(my $fh, '>', $STATE_FILE) or die "Kann $STATE_FILE nicht öffnen: $!";
print $fh to_json($state, {pretty => 1});
close($fh);
log_message("State gespeichert: $STATE_FILE");

File: javascript/sync_agent_run.js
#!/usr/bin/env node
// sync_agent_run.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:scripts/sync_agent_run.py
// auch in: OpenClaw@gateway2:scripts/sync_agent_run.py
// Erzeugt: 2026-08-19 durch ABSTRACTIONS_MANAGER.py

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

// Add custom module paths
const MODULE_PATHS = [
    '/home/openclaw/.openclaw/workspace/scripts'
];

// Simulate Python's sys.path.append
require('module').globalPaths.push(...MODULE_PATHS);

// Import required modules (these would need to be implemented in JS)
const { sync_to_git, sync_to_clawhub, log, validate_skill } = require('./sync_clawhub_git.js');

const CLAWHUB_DIR = '/home/openclaw/.openclaw/workspace/skills';
const GIT_DIR = '/home/openclaw/.openclaw/workspace/git/skills';

function file_mtime(dirPath) {
    try {
        const files = getAllFiles(dirPath).filter(file => !file.includes('.git'));
        if (files.length === 0) return 0;
        
        const mtimes = files.map(file => fs.statSync(file).mtime.getTime());
        return Math.max(...mtimes);
    } catch (error) {
        return 0;
    }
}

function getAllFiles(dirPath, arrayOfFiles = []) {
    const files = fs.readdirSync(dirPath);
    
    files.forEach(file => {
        const filePath = path.join(dirPath, file);
        if (fs.statSync(filePath).isDirectory()) {
            arrayOfFiles = getAllFiles(filePath, arrayOfFiles);
        } else {
            arrayOfFiles.push(filePath);
        }
    });
    
    return arrayOfFiles;
}

function getDirectories(source) {
    return fs.readdirSync(source, { withFileTypes: true })
        .filter(dirent => dirent.isDirectory() && !dirent.name.startsWith('.'))
        .map(dirent => dirent.name);
}

log("=".repeat(70));
log("CLAWHUB ↔ GIT SYNC AGENT - PRODUKTIONS-LAUF");
log(`Zeitstempel: ${new Date().toISOString()}`);
log("=".repeat(70));

const clawhub_skills = new Set(getDirectories(CLAWHUB_DIR));
const git_skills = new Set(getDirectories(GIT_DIR));

const results = {
    synced_to_git: [],
    synced_to_clawhub: [],
    up_to_date: [],
    errors: []
};

// 1. NEU in ClawHub → zu Git syncen
log("\n[PHASE 1] ClawHub → Git Synchronisation");
log("-".repeat(40));

const new_in_clawhub = [...clawhub_skills].filter(skill => !git_skills.has(skill)).sort();
for (const skill of new_in_clawhub) {
    try {
        if (validate_skill(path.join(CLAWHUB_DIR, skill))) {
            log(`→ Synchronisiere ${skill} zu Git...`);
            if (sync_to_git(skill, false)) {
                // Git init
                const git_path = path.join(GIT_DIR, skill);
                process.chdir(git_path);
                try {
                    execSync('git init -q 2>/dev/null', { stdio: 'ignore' });
                    execSync('git add . -f 2>/dev/null', { stdio: 'ignore' });
                    const dt = new Date().toLocaleString('sv-SE').slice(0, 16); // YYYY-MM-DD HH:mm
                    execSync(`git commit -m "Initial: ${skill}" -q 2>/dev/null`, { stdio: 'ignore' });
                } catch (e) {
                    // Ignore git errors
                }
                results.synced_to_git.push(skill);
                log(`  ✓ ${skill} synchronisiert & Git initialisiert`);
            } else {
                results.errors.push(`${skill} (sync failed)`);
            }
        } else {
            results.errors.push(`${skill} (invalid)`);
        }
    } catch (e) {
        log(`  ✗ ERROR: ${skill} - ${e.message}`, "ERROR");
        results.errors.push(`${skill} (exception)`);
    }
}

// 2. In beiden - prüfe Änderungen
log("\n[PHASE 2] Prüfe existierende Skills auf Änderungen");
log("-".repeat(40));

const in_both = [...clawhub_skills].filter(skill => git_skills.has(skill)).sort();
for (const skill of in_both) {
    try {
        const c_mtime = file_mtime(path.join(CLAWHUB_DIR, skill));
        const g_mtime = file_mtime(path.join(GIT_DIR, skill));
        const diff = (c_mtime - g_mtime) / 1000; // Convert to seconds

        if (Math.abs(diff) > 60) {
            if (diff > 0) {
                log(`→ ${skill}: ClawHub neuer (+${Math.round(diff)}s) → sync zu Git`);
                if (sync_to_git(skill, false)) {
                    const git_path = path.join(GIT_DIR, skill);
                    process.chdir(git_path);
                    try {
                        execSync('git add . -f 2>/dev/null', { stdio: 'ignore' });
                        const dt = new Date().toLocaleString('sv-SE').slice(0, 16);
                        execSync(`git commit -m "Sync from ClawHub: ${dt}" -q 2>/dev/null`, { stdio: 'ignore' });
                    } catch (e) {
                        // Ignore git errors
                    }
                    results.synced_to_git.push(skill);
                } else {
                    results.errors.push(`${skill} (update failed)`);
                }
            } else {
                log(`→ ${skill}: Git neuer (+${Math.round(Math.abs(diff))}s) → sync zu ClawHub`);
                if (sync_to_clawhub(skill, false)) {
                    results.synced_to_clawhub.push(skill);
                } else {
                    results.errors.push(`${skill} (update failed)`);
                }
            }
        } else {
            results.up_to_date.push(skill);
        }
    } catch (e) {
        log(`  ✗ ERROR: ${skill} - ${e.message}`, "ERROR");
        results.errors.push(`${skill} (exception)`);
    }
}

// ZUSAMMENFASSUNG
log("\n" + "=".repeat(70));
log("SYNCHRONISATION ABGESCHLOSSEN");
log("=".repeat(70));
log(`Zu Git synchronisiert:     ${results.synced_to_git.length}`);
if (results.synced_to_git.length > 0) {
    log(`  ${results.synced_to_git.join(', ')}`);
}
log(`Zu ClawHub synchronisiert: ${results.synced_to_clawhub.length}`);
if (results.synced_to_clawhub.length > 0) {
    log(`  ${results.synced_to_clawhub.join(', ')}`);
}
log(`Bereits aktuell:           ${results.up_to_date.length}`);
log(`Fehler:                    ${results.errors.length}`);
if (results.errors.length > 0) {
    log(`  ${results.errors.join(', ')}`);
}
log("=".repeat(70));

// Speichere State
const STATE_FILE = "/home/openclaw/.openclaw/workspace/db/sync_state.json";
const stateDir = path.dirname(STATE_FILE);
fs.mkdirSync(stateDir, { recursive: true });

const state = {
    last_run: new Date().toISOString(),
    results: results
};

fs.writeFileSync(STATE_FILE, JSON.stringify(state, null, 2));
log(`State gespeichert: ${STATE_FILE}`);
